### PR TITLE
APPS/IO-DEMO: Cancel connection establishment if disconnecting

### PR DESCRIPTION
**Pull request by dmitrygx**
backporting of https://github.com/openucx/ucx/pull/7216

### CODE REVIEW

**Comment by dmitrygx** [comment]
this PR fixes segfaults and assertions as on the master branch.
no core files were found after running io-demo run script with this PR:
/hpc/mtr_scrap/users/dmitrygla/io_demo/timeout_waiting_for_replies_dc/repro_1628543161-462959540/1629488017-724651992

**Comment by yosefe** [comment]
@dmitrygx is this PR up-to-date with master? were there any changes conflicts?

**Comment by dmitrygx** [comment]
@yosefe 

> is this PR up-to-date with master?

yes, it is aligned with the latest state of #7216 which has already been merged to the master

> were there any changes conflicts?

here is the list of conflicts:
```diff
  void UcxConnection::disconnect(UcxCallback *callback)
  {
-     /* establish cb must be destroyed earlier since it accesses
-      * the connection */
-     assert(_establish_cb == NULL);
+     UCX_CONN_LOG << "disconnect, ep is " << _ep;
+
      assert(_disconnect_cb == NULL);
+     _disconnect_cb = callback;

++<<<<<<< HEAD
 +    UCX_CONN_LOG << "destroying, ep is " << _ep;
 +
 +    _disconnect_cb = callback;
 +    if (ucs_list_is_empty(&_all_requests)) {
 +        ep_close(UCP_EP_CLOSE_MODE_FORCE);
 +        _context.move_connection_to_disconnecting(this);
 +    } else {
 +        cancel_all();
 +        ep_close(UCP_EP_CLOSE_MODE_FORCE);
++=======
+     _context.remove_connection(this);
+
+     if (!is_established()) {
+         assert(_ucx_status == UCS_INPROGRESS);
+         established(UCS_ERR_CANCELED);
++>>>>>>> 96ad72e... APPS/IO-DEMO: Cancel connection establishment if disconnecting, destroy connections in UсxContext dtor
      }
+
+     _context.move_connection_to_disconnecting(this);
+
+     cancel_all();
+
+     // close the EP after cancelling all outstanding operations to purge all
+     // requests scheduled on the EP which could wait for the acknowledgments
+     ep_close(UCP_EP_CLOSE_MODE_FORCE);
  }
```
```diff
  bool UcxConnection::disconnect_progress()
@@@ -718,7 -852,8 +772,12 @@@
          }
      }

++<<<<<<< HEAD
 +    assert(ucs_list_is_empty(&_all_requests));
++=======
+     UCX_CONN_LOG << "disconnection completed";
+
++>>>>>>> 96ad72e... APPS/IO-DEMO: Cancel connection establishment if disconnecting, destroy connections in UсxContext dtor
      invoke_callback(_disconnect_cb, UCS_OK);
      return true;
  }
```
```diff
@@@ -974,13 -1176,10 +1041,20 @@@ void UcxConnection::request_completed(u
      ucs_list_del(&r->pos);

      if (_disconnect_cb != NULL) {
++<<<<<<< HEAD
 +        UCX_CONN_LOG << "completing request " << r << " with status \""
 +                     << ucs_status_string(r->status) << "\" (" << r->status
 +                     << ")" << " during disconnect";
 +
 +        if (ucs_list_is_empty(&_all_requests)) {
 +            _context.move_connection_to_disconnecting(this);
 +        }
++=======
+         UCX_CONN_LOG << "completing " << r->what << " request " << r
+                      << " with status \"" << ucs_status_string(r->status)
+                      << "\" (" << r->status << ")" << " during disconnect";
+         assert(_context.is_in_disconnecting_list(this));
++>>>>>>> 96ad72e... APPS/IO-DEMO: Cancel connection establishment if disconnecting, destroy connections in UсxContext dtor
      }
  }
```
```diff
@@@ -123,10 -111,6 +124,13 @@@ public
      static const std::string sockaddr_str(const struct sockaddr* saddr,
                                            size_t addrlen);

++<<<<<<< HEAD
 +    void destroy_connections();
 +
 +    static double get_time(const struct timeval &tv);
 +
++=======
++>>>>>>> 96ad72e... APPS/IO-DEMO: Cancel connection establishment if disconnecting, destroy connections in UсxContext dtor
      static double get_time();

      static void *malloc(size_t size, const char *name);
```

File: test/apps/iodemo/ucx_wrapper.cc [hunk 1]
```diff
@@ -687,28 +722,35 @@ void UcxConnection::accept(ucp_conn_request_h conn_req, UcxCallback *callback)
...
+        established(UCS_ERR_CANCELED);
     }
+
+    _context.move_connection_to_disconnecting(this);
+
+    cancel_all();
+
+    // close the EP after cancelling all outstanding operations to purge all
+    // requests scheduled on the EP which could wait for the acknowledgments
+    ep_close(UCP_EP_CLOSE_MODE_FORCE);
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:743]: but we should not close the endpoint until request list is empty on int3 branch
I think need to move this back to disconnect_progress

Comment (dmitrygx) [reply test/apps/iodemo/ucx_wrapper.cc:743]: done

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@dmitrygx can you pls make sure this actually works with int3 io_demo (in case we missed something) before merging?

**Comment by dmitrygx** [comment]
> @dmitrygx can you pls make sure this actually works with int3 io_demo (in case we missed something) before merging?

@yosefe I tested and it works fine: no core files, hangs were found during running the test which reproduces “timeout waiting for <N> replies”.
But I think we should run the test from our AZP CI which was able to reproduce issue with the mater branch. I’ll update herewhne completed

**Comment by dmitrygx** [comment]
@yosefe it hangs when killing the client process, it can't destroy the connection, requests aren't cancelled.
but if I remove the second commit, it works fine

**Comment by dmitrygx** [comment]
I checked as discussed offline, the client process hangs
```
...
[1629718513.762455] [DEMO] not disconnecting [UCX-connection 0x1288e60: #2 2.1.4.3:40000] read 0/0 write 0/0 (disconnecting) due to "End of the Client run" because disconnection is already in progress
[1629718560.375233] [sputnik4:10044:0]    ib_mlx5_log.c:143  UCX  ERROR Transport retry count exceeded on mlx5_bond_0:1/RoCE (synd 0x15 vend 0x81 hw_synd 0/0)
[1629718560.375233] [sputnik4:10044:0]    ib_mlx5_log.c:143  UCX  ERROR RC QP 0x2e08b wqe[3]: SEND --- [inl len 2]
```
attaching to the processes shows that `_all_requests` list isn't empty even doing `cancel_all()` for all outstanding recv/send requests on int3
```
$ gdb -p 10044
...
(gdb) bt
#0  0x00007fd22f96a6fb in ucs_recursive_spin_lock (lock=0x1193ce0) at /labhome/dmitrygla/ucx/src/ucs/type/spinlock.h:120
#1  0x00007fd22f96a9de in ucs_callbackq_enter (cbq=0x1193c20) at datastruct/callbackq.c:62
#2  0x00007fd22f96ba6f in ucs_callbackq_slow_proxy (arg=0x1193c20) at datastruct/callbackq.c:374
#3  0x00007fd22f646c99 in ucs_callbackq_dispatch (cbq=0x1193c20) at /labhome/dmitrygla/ucx/src/ucs/datastruct/callbackq.h:211
#4  0x00007fd22f64d95d in uct_worker_progress (worker=0x1193c20) at /labhome/dmitrygla/ucx/src/uct/api/uct.h:2374
#5  ucp_worker_progress (worker=0x1199010) at core/ucp_worker.c:1970
#6  0x00000000004053ba in UcxContext::wait_disconnected_connections (this=0x7ffeafc70d70) at ucx_wrapper.cc:544
#7  0x0000000000413f16 in DemoClient::destroy_servers (this=0x7ffeafc70d70) at io_demo.cc:1517
#8  0x0000000000414578 in DemoClient::run (this=0x7ffeafc70d70) at io_demo.cc:1618
#9  0x000000000040e1af in do_client (test_opts=...) at io_demo.cc:2151
#10 0x000000000040e5a9 in main (argc=14, argv=0x7ffeafc71508) at io_demo.cc:2194
(gdb) n
...
UcxConnection::disconnect_progress (this=0x1288e60) at ucx_wrapper.cc:744
744         assert(_disconnect_cb != NULL);
(gdb) n
746         if (!ucs_list_is_empty(&_all_requests)) {
(gdb)
747             return false;
(gdb)
...
```

**Comment by dmitrygx** [comment]
so, it seems that doing `ep_close()` to cancel all requests is required for in3 too

**Comment by yosefe** [comment]
@dmitrygx you are right, i looked again on the previous code, and we always did ep_close after cancel_all immediately.
so let's revert the 2nd commit

**Comment by dmitrygx** [comment]
> @dmitrygx you are right, i looked again on the previous code, and we always did ep_close after cancel_all immediately.
> so let's revert the 2nd commit

@yosefe done

**Review by yosefe — APPROVED** [review]

**Pull request merged**